test/unit/gateways/blue_snap_test.rb

```diff
@@ -55,6 +65,37 @@ def test_successful_echeck_purchase
     assert_equal '1019803029', response.authorization
   end
 
+  def test_successful_purchase_with_3ds_auth
+    response = stub_comms(@gateway, :raw_ssl_request) do
+      @gateway.purchase(@amount, @credit_card, @options_3ds2)
+    end.check_request do |method, url, data|
+      assert_match(/<three-d-secure>/, data)
+      assert_match(/<eci>#{Regexp.quote(@options_3ds2[:three_d_secure][:eci])}<\/eci>/, data)
+      assert_match(/<cavv>#{Regexp.quote(@options_3ds2[:three_d_secure][:cavv])}<\/cavv>/, data)
+      assert_match(/<xid>#{Regexp.quote(@options_3ds2[:three_d_secure][:xid])}<\/xid>/, data)
+      assert_match(/<three-d-secure-version>#{Regexp.quote(@options_3ds2[:three_d_secure][:version])}<\/three-d-secure-version>/, data)
+      assert_match(/<ds-transaction-id>#{Regexp.quote(@options_3ds2[:three_d_secure][:ds_transaction_id])}<\/ds-transaction-id>/, data)
+    end.respond_with(successful_purchase_with_3ds_auth_response)
+
+    assert_success response
+    assert_equal '1024951831', response.authorization
+    assert_equal '019082915501456', response.params['original-network-transaction-id']
+    assert_equal '019082915501456', response.params['network-transaction-id']
```

Comment:
I've observed that the network transaction id values are included in the responses to the purchase and auth requests which include third-party 3DS auth data. These fields are duplicated in different places in the response XML document that is returned, but they seem to be referring to the same value.

Since our current response parsing strategy flattens out the returned XML document, _if_ the fields with the same names but at different paths ever did need to be distinguished, we will need to do more work. Decided it would be better to wait until we find we need to do that extra work rather than try to preemptively do it now.

Reply:
Will there be a situation that they would have different values? If they are referring to the same value, I agree that it doesn't seem like a problem.

Reply:
@molbrown Yah, that was sort of bothering me, so I just dug a little deeper into BlueSnap's documentation to verify. It does appear to be the case that `network-transaction-id` is just duplicated in two different places in the XML response. From [their documentation](https://developers.bluesnap.com/v8976-JSON/docs/auth-capture#section--3-d-secure-examples-):

>### 3-D Secure Examples
>#### 3DS with shopper-initiated transaction
>
>The request includes `"transactionInitiator":"SHOPPER"` to indicate that this is a Customer-Initiated Transaction. The response includes the same `networkTransactionId` in both the `networkTransactionInfo` and the `processingInfo` sections. The merchant keeps this `networkTransactionId` for future transactions.
>
>#### 3DS with MIT
>
>The request includes `"transactionInitiator":"MERCHANT"` to indicate that this is a Merchant-Initiated Transaction. The merchant also includes the `originalNetworkTransactionId` to link the transaction to an existing transaction on which SCA was performed so SCA does not need to be performed again. In the `networkTransactionInfo` section, the response includes both the `originalNetworkTransactionId` and the new `networkTransactionId`. The merchant continues using the `originalNetworkTransactionId` for future transactions.

So I think we're good to go on this point. One more thing that occurs to me is: we may soon wish to implement the Stored Credentials framework for BlueSnap and as part of that ensure that original network transaction id is being returned with the AM response for follow-up requests.